components/terms_of_service/terms_of_service.jsx

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import {Button, ButtonGroup} from 'react-bootstrap';
+import {FormattedMessage} from 'react-intl';
+
+import {memoizeResult} from 'mattermost-redux/utils/helpers';
+
+import * as GlobalActions from 'actions/global_actions.jsx';
+import {getServiceTerms, updateServiceTermsStatus} from 'actions/user_actions.jsx';
+
+import AnnouncementBar from 'components/announcement_bar';
+import FormattedMarkdownMessage from 'components/formatted_markdown_message.jsx';
+import LoadingScreen from 'components/loading_screen.jsx';
+
+import {browserHistory} from 'utils/browser_history';
+import messageHtmlToComponent from 'utils/message_html_to_component';
+import {formatText} from 'utils/text_formatting.jsx';
+import * as Utils from 'utils/utils.jsx';
+import {Constants} from 'utils/constants.jsx';
+
+export default class TermsOfService extends React.PureComponent {
+    static propTypes = {
+        customServiceTermsId: PropTypes.string.isRequired,
+        privacyPolicyLink: PropTypes.string,
+        siteName: PropTypes.string,
+        termsEnabled: PropTypes.bool,
+        termsOfServiceLink: PropTypes.string,
+    };
+
+    static defaultProps = {
+        privacyPolicyLink: 'https://about.mattermost.com/default-privacy-policy/',
+        siteName: 'Mattermost',
+        termsEnabled: true,
+        termsOfServiceLink: 'https://about.mattermost.com/default-terms/',
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            customServiceTermsText: '',
+            loading: true,
+            loadingAgree: false,
+            loadingDisagree: false,
+            serverError: null,
+        };
+
+        this.formattedText = memoizeResult((text) => formatText(text));
+    }
+
+    componentDidMount() {
+        if (this.props.termsEnabled) {
+            this.getServiceTerms();
+        } else {
+            GlobalActions.redirectUserToDefaultTeam();
+        }
+    }
+
+    componentDidUpdate(prevProps) {
+        if (prevProps.customServiceTermsId !== this.props.customServiceTermsId) {
+            this.getServiceTerms();
+        }
+    }
+
+    getServiceTerms = () => {
+        this.setState({
+            customServiceTermsText: '',
+            loading: false,
+        });
+        getServiceTerms(
+            (data) => {
+                this.setState({
+                    customServiceTermsText: data.text,
+                    loading: false,
+                });
+            },
+            () => {
+                GlobalActions.emitUserLoggedOutEvent(`/login?extra=${Constants.GET_TERMS_ERROR}`);
```

Comment:
Is this the right response to a server failure to fetch the service terms? I'm wondering if it makes more sense to leave them on this page and just display the error.

Reply:
The text would not be loaded and we would need to hide the `Agree` and `Disagree` buttons and the footer text. Not much will be left on the page.
Thus, this response was proposed by @jasonblais [here](https://pre-release.mattermost.com/core/pl/93pt9f5xyid1ifuxqted66wikr).

Reply:
Seems reasonable!